### DISCUSSION TITLE
XML Data in Get Request

### DISCUSSION THREAD

**Discussion by smurraystrata** [Budibase/budibase#13176] [Help]
Hello again all!

I have a service I can make basic HTTP get requests to, which responds with XML. I'm trying to figure out how to access that day, or transform it into JSON so it works with Budibase's tables. I have tried using DOMParser, but I get Query Error: DOMParser is not defined. I tried basic transformers ([like this](https://github.com/Budibase/budibase/discussions/9758)) but I do not get data when I try it. Any suggestions?

Here's a small snippet from the beginning of the RAW and JSON responses. I'm not putting the whole response as it is quite large and would take a while to remove data from... Let me know if more is needed.

`<?xml version="1.0"?>
<ranges>
  <dhcpv4Ranges>
    <range>
      <name>name</name>
      <description/>
      <startip>1.1.1.1.</startip>
      <endip>2.2.2.2</endip>
  	  <subnetmask>255.255.255.255</subnetmask>
      <criteria>(CLIENTCLASS(class))</criteria>
      <aggregateCriteria>(NOT DOCSIS(*) AND NOT PKTC[*]) AND ((CLIENTCLASS(class)))</aggregateCriteria>`

JSON:
`{
  "value": "<?xml version=\"1.0\"?>\r\n<ranges>\r\n  <dhcpv4Ranges>\r\n    <range>\r\n      <name>name</name>\r\n      <description/>\r\n      <startip>1.1.1.1</startip>\r\n      <endip>2.2.2.2</endip>\r\n  \t  <subnetmask>255.255.255.255</subnetmask>\r\n      <criteria>(CLIENTCLASS(class))</criteria>\r\n      <aggregateCriteria>(NOT DOCSIS(*) AND NOT PKTC[*]) AND ((CLIENTCLASS(class)))</aggregateCriteria>\r\n `

Here's one thing I've tried for the transformer, which I thought would work based on the linked discussion above, but I get {} when I use it:

`return data.dhcpv4Ranges`

Thank you for this great tool and the help I've had previously!

**Comment by ConorWebb96** [comment]
Hey @jedman20,

You should be able to use your xml output directly. As you only provided a single item, I'll only be showing how to get the fields listed in your above example. In my case I've opted to output the startip, as it was the first field that had a value.

Rest request output + transformer
<img width="968" alt="Screenshot 2024-03-05 at 08 59 13" src="https://github.com/Budibase/budibase/assets/126772285/07a607dc-520d-46d0-8c72-44160dfd9cdc">

In the design view i've used these components 
- Data provider
<img width="1509" alt="Screenshot 2024-03-05 at 09 03 33" src="https://github.com/Budibase/budibase/assets/126772285/4f2c9ad6-4a3a-4887-a3c6-8365f0e8896a">
- Repeater
<img width="1512" alt="Screenshot 2024-03-05 at 09 03 46" src="https://github.com/Budibase/budibase/assets/126772285/12db43aa-f116-4c42-8e6b-0470cb875e9d">
- Paragraph - code inside the paragraph can be found below.

```js
const ranges = $("ranges.Rows");
return ranges[$("New Repeater.Row index")].range.startip
```

As you can seen in the screenshots it outputs the value stored in the field. If you wanted to get the other fields, you'd simply switch out startip with the other field names.

I hope this helps.

**Reply by smurraystrata** [reply]
Hello @ConorWebb96,
Thank you for the effort you put into your reply, I must be doing something wrong though. 

I tried both data.ranges and data.dhcpv4Ranges - I get this for the JSON:
![image](https://github.com/Budibase/budibase/assets/120056389/f7dcd862-da70-4d66-9a92-902ede73e81d)

And this for RAW:
![image](https://github.com/Budibase/budibase/assets/120056389/1bb11295-3887-44af-ad23-0fec77876606)

I did notice this response does not have a content type header, could that be part of the problem?
![image](https://github.com/Budibase/budibase/assets/120056389/498ec85d-b0fd-4494-8268-8c13552a191e)

Thank you again for the help!

**Reply by ConorWebb96** [reply]
- Could this be related to the rest api itself that you are using?
- What happens if you just output the `data` in the transformer? (Is anything shown? What does it show in JSON?)

What happens when you try to output the data in something like Postman? 

The content type attached to the rest request I demoed was `text/xml` with `utf-8` encoding. Hopefully this is helpful.

I'm not sure if there a lot more I can help with, as I believe it likely related to how your API is setup.

**Reply by smurraystrata** [reply]
Could this be related to the rest api itself that you are using?

- I'm sure it is now. Until I was typing my last response I hadn't noticed the lack of a content-type header.

What happens if you just output the data in the transformer? (Is anything shown? What does it show in JSON?

- I get the key 'value' with the entire XML response listed as the value. I haven't been able to break the pieces out of the response.
![image](https://github.com/Budibase/budibase/assets/120056389/184c5fc0-1121-45ed-86f9-bc87c2fe3ed8)

Postman does correctly parse it into JSON, not sure what magic it has that makes that work. I think you are correct it is a problem with how the API is set up. I was hoping it was a simple pebkac error though :D. Thank you for looking into this!
